appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/extension/cluster/ClusterScopedInterceptor.java

```diff
@@ -0,0 +1,151 @@
+ * file that accompanied this code.
+ *
+ * Modifications:
+ * If applicable, add the following below the License Header, with the fields
+ * enclosed by brackets [] replaced by your own identifying information:
+ * "Portions Copyright [year] [name of copyright owner]"
+ *
+ * Contributor(s):
+ * If you wish your version of this file to be governed by only the CDDL or
+ * only the GPL Version 2, indicate your decision by adding "[Contributor]
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package fish.payara.micro.cdi.extension.cluster;
+
+import com.google.common.collect.Iterables;
+import com.hazelcast.core.IAtomicLong;
+import fish.payara.cluster.Clustered;
+import fish.payara.cluster.DistributedLockType;
+import static fish.payara.micro.cdi.extension.cluster.ClusterScopeContext.getAnnotation;
+import static fish.payara.micro.cdi.extension.cluster.ClusterScopeContext.getBeanName;
+import fish.payara.micro.cdi.extension.cluster.annotations.ClusterScoped;
+import fish.payara.micro.cdi.extension.cluster.annotations.ClusterScopedIntercepted;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.Serializable;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.annotation.Priority;
+import javax.enterprise.context.spi.Context;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.CDI;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+import lombok.extern.java.Log;
+import org.glassfish.api.invocation.InvocationManager;
+import org.glassfish.internal.api.Globals;
+
+/**
+ * Intercepts every method call to refresh the cluster
+ *
+ * @author lprimak
+ */
+@Interceptor @ClusterScopedIntercepted @Log @Priority(Interceptor.Priority.PLATFORM_AFTER)
+public class ClusterScopedInterceptor implements Serializable {
+    private final BeanManager bm = CDI.current().getBeanManager();
+    private transient ClusteredSingletonLookupImpl clusteredLookup;
+    private static final long serialVersionUID = 1L;
+
+
+    public ClusterScopedInterceptor() {
+        init();
+    }
+
+    @AroundInvoke
+    public Object lockAndRefresh(InvocationContext invocationContext) throws Exception {
+        Class<?> beanClass = invocationContext.getMethod().getDeclaringClass();
+        Clustered clusteredAnnotation = beanClass.getAnnotation(Clustered.class);
```

Comment:
Note this will not work when stereotypes are used, or when the annotation is added by an extension or factory. See the RolesPermitted Interceptor or the Soteria Interceptors.